### PR TITLE
Add kube support for availability zones

### PR DESCRIPTION
**Pull request by liamawhite**
Solves #737

### CODE REVIEW

File: platform/kube/cache.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	ch.handler.Append(func(obj interface{}, ev model.Event) error {
+		node := *obj.(*v1.Node)
+		var ip string
+		for _, a := range node.Status.Addresses {
+			if a.Type == "InternalIP" {
+				ip = a.Address
+				break
+			}
+		}
+		if len(ip) > 0 {
```
Comment (rkpagadala) [inline platform/kube/cache.go]: @rshriram the code is correct need > 0.

File: platform/kube/cache.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+				delete(out.keys, ip)
+			}
+		}
+		return nil
+	})
+	return out
+}
+
+// getNodeByIP returns the node or nil if pod not found or an error occurred
+func (nc *NodeCache) getNodeByIP(addr string) (*v1.Node, bool) {
```
Comment (kyessenov) [inline platform/kube/cache.go]: That's a common pattern in go. It's not expensive to return multiple values.

File: platform/kube/cache.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	item, exists, err := nc.informer.GetStore().GetByKey(key)
+	if !exists || err != nil {
+		return nil, false
+	}
+	return item.(*v1.Node), true
+}
+
+// getAZByIP returns the az and true if found
+// empty string and false if the node couldn't be found or an error occurred
+func (nc *NodeCache) getAZByIP(addr string) (string, bool) {
```
Comment (kyessenov) [inline platform/kube/cache.go]: It lets us avoid assigning special meaning to empty string. Common pattern.

File: platform/kube/controller.go [hunk 4]
```diff
@@ -33,6 +33,9 @@ import (
 	"istio.io/pilot/model"
 )
 
+// NodeAZLabel is the well-known label for kubernetes node AZs
```
Comment (kyessenov) [inline platform/kube/controller.go]: Yes, please correct the doc.

File: platform/kube/cache.go (hunk 2 above)
Comment (liamawhite) [inline platform/kube/cache.go]: Yeah, that and I wanted to keep it in a similar structure to podcache. Also exists/bool often makes code more readable.

File: platform/kube/controller.go (hunk 4 above)
Comment (liamawhite) [inline platform/kube/controller.go]: What's the difference? I went with labels based on the definitions of the two in the docs.

File: platform/kube/cache.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		return nil, false
+	}
+	return convertTags(pod.ObjectMeta), true
+}
+
+// NodeCache is an eventually consistent node cache
+type NodeCache struct {
+	cacheHandler
+
+	// keys maintains stable node internal IP to name key mapping
```
Comment (liamawhite) [inline platform/kube/cache.go]: I'm not 100% sure of the differences when in kube? Presumably its internal/external to the cluster?

File: platform/kube/controller.go (hunk 4 above)
Comment (liamawhite) [inline platform/kube/controller.go]:
```
metadata:
  annotations:
    volumes.kubernetes.io/controller-managed-attach-detach: "true"
  creationTimestamp: 2017-07-19T19:39:27Z
  labels:
    arch: amd64
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: us-south
    failure-domain.beta.kubernetes.io/zone: hou02
``` 
Are we sure its not a label? 😕

**Comment by liamawhite** [comment]
@rshriram @kyessenov Just added some more tests, should be ready to for full review. Implementation has changed slightly due to the ability to use node names as pointed out by @esnible, see code comments for link to kube docs that describe the relevant behaviour.

As discussed on istio-dev group this is turned on by default but a further issue will be created to enable users to set the `upstream.zone_routing.min_cluster_size` of envoy as documented [here](https://lyft.github.io/envoy/docs/configuration/cluster_manager/cluster_runtime.html#zone-aware-load-balancing). This needs a separate issue as its not a trivial piece of work.

**Comment by liamawhite** [comment]
Need to write a few more tests in `proxy/envoy`...

File: platform/kube/controller.go (hunk 4 above)
Comment (rshriram) [inline platform/kube/controller.go]: Fine, lets stick with label. :)

**Comment by liamawhite** [comment]
Not sure how code coverage has decreased, I wrote a whole bunch of new tests for code that was already there.

**Comment by rshriram** [comment]
@istio/pilot-hackers - please do the Go code review.

File: platform/kube/controller.go [hunk 6]
```diff
@@ -33,6 +33,9 @@ import (
...
+const NodeAZLabel = "failure-domain.beta.kubernetes.io/zone"
+
```
Comment (rshriram) [inline platform/kube/controller.go]: Don't you want to process the region as well?

Comment (rshriram) [inline platform/kube/controller.go]: Essentially you can flatten region and zone into one string since Envoy does not have support for regions

Comment (liamawhite) [inline platform/kube/controller.go]: Done!

File: platform/kube/controller.go [hunk 7]
```diff
@@ -33,6 +33,13 @@ import (
 	"istio.io/pilot/model"
 )
 
+const (
+	// NodeRegionLabel is the well-known label for kubernetes node region
+	NodeRegionLabel = "failure-domain.beta.kubernetes.io/region"
+	// NodeZoneLabel is the well-known label for kubernetes node zone
+	NodeZoneLabel = "failure-domain.beta.kubernetes.io/zone"
```
Comment (kyessenov) [inline platform/kube/controller.go]: @rshriram the examples show that the zone includes the region.

**Pull request merged**